### ISSUE TITLE
ddp doc.

### ISSUE THREAD

**Issue by sbelharbi** [pytorch/pytorch#64150]
## 📚 Documentation

hi,
this is about the doc of ddp (DistributedDataParallel). i may be wrong, but here are two related points, 
1. i couldnt find a `complete` tutorial on using ddp, in particular examples that use dataloader with multiple workers. i found these pages: [p1](https://pytorch.org/tutorials/intermediate/ddp_tutorial.html), [p2](https://pytorch.org/docs/stable/notes/ddp.html), [p3](https://pytorch.org/tutorials/beginner/dist_overview.html), [p4](https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html#torch.nn.parallel.DistributedDataParallel), [p5](https://pytorch.org/tutorials/intermediate/dist_tuto.html), [p6](https://pytorch.org/docs/stable/distributed.html#distributed-basics), [p7](https://github.com/ml-edu/examples/blob/master/imagenet/main.py), [p8](https://github.com/pytorch/examples/blob/master/distributed/ddp/README.md). while all these pages provide very helpful information, they often discuss a specific part of ddp or they sem to use a different way to kick start processes. including dataloaders with multilple works, access to model attributes for a model tha has been wrapped within ddp could be very helpful to quickly start using this module.
2. the second point is about the difference between all the links about on starting multiprocessing. there may be different ways to start multiprocess in ddp (here, process is the one that will run on different gpu). it seems like some pages using `torch.multiprocessing`, then expliclty call `spawn` to create multiple process combined with `torch.distributed.init_process_group`. other, simply call `torch.distrbuted.init_process_group` without `torch.multiprocessing.spawn`. for example, this [page](https://github.com/pytorch/examples/blob/master/distributed/ddp/README.md) uses the last method . so, it is not clear about the `spawn` part. i may be wrong, but probably this is most dated way to use ddp. the last method seems to use the pytorch [launcher](https://pytorch.org/docs/stable/distributed.html) that automatically spawns. now after reading again the launcher part, it seems as a second option to run distributed computations that is helpful in clusters with multiples nodes. this brings me to the last part that is related to spawn, dataloader and multiple workers. 
3. i found this warning in the [doc](https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html#torch.nn.parallel.DistributedDataParallel) of ddp `If you plan on using this module with a nccl backend or a gloo backend (that uses Infiniband), together with a DataLoader that uses multiple workers, please change the multiprocessing start method to forkserver (Python 3 only) or spawn. Unfortunately Gloo (that uses Infiniband) and NCCL2 are not fork safe, and you will likely experience deadlocks if you don’t change this setting.`. because ddp can be used in a transparent way without calling spawn, and because when using dataloader with multiple workers we dont specify the forking method, it is not clear where we should set this forking method. the method reference here `set_start_method` is in `torch.multiprocessing`, since when using ddp, it does not seem necessary to use `torch.multiprocessing` unless you intentionally want to create furthermore process other the ones created with `torch.distributed` to run on each gpu. 
 i guess if i use the launcher, i shouldnt worry about the forking method because it is `torch.multiprocessing.spawn` that has an argument named `start_method` which refers to start_processes() to use a different method other than `spawn` but it did not mention `forkserver`. setting the start method could have something to do with `multiprocessing_context` in [dataloader](https://pytorch.org/docs/stable/data.html#torch.utils.data.DataLoader) which is by the way undocumented. from what i understood from examples, using multiple workers with dataloader is transparent when using ddp. for example, without ddp, asking to use `n` workers will create `n` processes to load the data. the same thing is expected to happen when using ddp but for each `ddp process`. so far i am not sure about this because of the above warning. also, now that dataloader is combined with a distributed [sampler](https://pytorch.org/docs/stable/data.html). 

in short, it would be very helpful to illustrate a complete way to use ddps.
sorry for the long text.
hope this will improve the doc.
thanks

**Comment by zhaojuanmao** [comment]
@soulitzer  good point to add an complete example  for distributed training, including data loading, ddp construction, training loop and etc.  

regarding some of your questions: 
1.  if you use launcher, you do not need to call process spawn, as launcher will do that for you; otherwise, you need to call process spawn in your codes to launch processes 
2.  data loading codes are usually written before training loop starts in each process
3.  if you want to evaluate loss from each rank in master, you can all dist.all_gather()

**Comment by sbelharbi** [comment]
yes, thanks
i do use launch. and it seems we need to switch to [dist.run](https://github.com/pytorch/pytorch/blob/65e6194aeb3269a182cfe2c05c122159da12770f/torch/distributed/run.py) now...

and for the sync, yes i am using now dist.all_gather.
to synch a tensor of some values:

```python
def sync_tensor_across_gpus(t: Union[torch.Tensor, None]
                            ) -> Union[torch.Tensor, None]:
    if t is None:
        return None
    group = dist.group.WORLD
    group_size = torch.distributed.get_world_size(group)
    gather_t_tensor = [torch.zeros_like(t) for _ in
                       range(group_size)]
    dist.all_gather(gather_t_tensor, t)
    return torch.cat(gather_t_tensor, dim=0)


# ...
tensor = compute_stuff()
# want to sync tensor across all processes:
tensor = sync_tensor_across_gpus(tensor)  # can do different ops over the gathered tensor: sum, average, ...
```

thanks
